test_values/cached_proof_tests.json

```diff
@@ -0,0 +1,41 @@
+[
+    {
+        "update": {
+            "adds": [20, 21, 22],
+            "proof": {
+                "targets": [0, 1],
+                "hashes": [
+                    "9576f4ade6e9bc3a6458b506ce3e4e890df29cb14cb5d3d887672aef55647a2b",
+                    "29590a14c1b09384b94a2c0e94bf821ca75b62eacebc47893397ca88e3bbcbd7"
+                ]
+            },
+            "del_hashes": [
+                "6e340b9cffb37a989ca544e6bb780a2c78901d3fb33738768511a30617afa01d",
+                "4bf5122f344554c53bde2ebb8cd2b7e3d1600ad631c385a5d7cce23c7785459a"
+            ]
+        },
+        "cached_proof": {
+            "targets": [4, 6, 7],
+            "hashes": [
+                "e77b9a9ae9e30b0dbdb6f510a264ef9de781501d7b6b92ae89eb059c5ab743db",
+                "df46b17be5f66f0750a4b3efa26d4679db170a72d41eb56c3e4ff75a58c65386"
+            ]
+        },
+        "initial_stump": "Add 8 leaves [0, 1, 2, 3, 4, 5, 6, 7]",
+        "initial_roots": [
+            "b151a956139bb821d4effa34ea95c17560e0135d1e4661fc23cedc3af49dac42"
+        ],
+        "initial_leaves": 8,
+        "cached_hashes": [
+            "e52d9c508c502347344d8c07ad91cbd6068afc75ff6292f062a09ca381c89e71",
+            "67586e98fad27da0b9968bc039a1ef34c939b9b8e523a8bef89d478608c5ecf6",
+            "ca358758f6d27e6cf45272937977a748fd88391db679ceda7dc7bf1f005ee879"
+        ],
+        "expected_roots": [
+            "97491b30a42410dc3267d17933cf5e1b55cfb92ebab2dcf1bcd098032dacee95",
+            "0d451c2d9366705a3557fd038c25a40a348cc0effa2dd00dfaaba65749cd0915",
+            "7cb7c4547cf2653590d7a9ace60cc623d25148adfbc88a89aeb0ef88da7839ba"
+        ],
+        "expected_targets": [4, 6, 7]
+    }
+]
```

Comment:
The below test case should pass:

```
    {
        "update": {
            "adds": [8],
            "proof": {
                "targets": [0, 1, 4],
                "hashes": [
                    "e77b9a9ae9e30b0dbdb6f510a264ef9de781501d7b6b92ae89eb059c5ab743db",
                    "9576f4ade6e9bc3a6458b506ce3e4e890df29cb14cb5d3d887672aef55647a2b",
                    "34028bbc87000c39476cdc60cf80ca32d579b3a0e2d3f80e0ad8c3739a01aa91"
                ]
            },
            "del_hashes": [
                "6e340b9cffb37a989ca544e6bb780a2c78901d3fb33738768511a30617afa01d",
                "4bf5122f344554c53bde2ebb8cd2b7e3d1600ad631c385a5d7cce23c7785459a",
                "e52d9c508c502347344d8c07ad91cbd6068afc75ff6292f062a09ca381c89e71"
            ]
        },
        "remembers": [0],
        "cached_proof": {
            "targets": [4, 6, 7],
            "hashes": [
                "e77b9a9ae9e30b0dbdb6f510a264ef9de781501d7b6b92ae89eb059c5ab743db",
                "df46b17be5f66f0750a4b3efa26d4679db170a72d41eb56c3e4ff75a58c65386"
            ]
        },
        "initial_stump": "Add 8 leaves [0, 1, 2, 3, 4, 5, 6, 7]",
        "initial_roots": [
            "b151a956139bb821d4effa34ea95c17560e0135d1e4661fc23cedc3af49dac42"
        ],
        "initial_leaves": 8,
        "cached_hashes": [
            "e52d9c508c502347344d8c07ad91cbd6068afc75ff6292f062a09ca381c89e71",
            "67586e98fad27da0b9968bc039a1ef34c939b9b8e523a8bef89d478608c5ecf6",
            "ca358758f6d27e6cf45272937977a748fd88391db679ceda7dc7bf1f005ee879"
        ],
        "expected_roots": [
            "7f6df6757821144278f5d535261ab4ffc55e2618f7bc62f1a19e7d289061b08b",
            "6e340b9cffb37a989ca544e6bb780a2c78901d3fb33738768511a30617afa01d"
        ],
        "expected_targets": [6, 7, 8],
        "expected_cached_hashes": [
                "67586e98fad27da0b9968bc039a1ef34c939b9b8e523a8bef89d478608c5ecf6",
                "ca358758f6d27e6cf45272937977a748fd88391db679ceda7dc7bf1f005ee879",
                "6e340b9cffb37a989ca544e6bb780a2c78901d3fb33738768511a30617afa01d"
        ]
    }
```

diff of the `cached_proof_tests.json`:
```diff
diff --git a/test_values/cached_proof_tests.json b/test_values/cached_proof_tests.json
index 6b27f11..bf4e1c2 100644
--- a/test_values/cached_proof_tests.json
+++ b/test_values/cached_proof_tests.json
@@ -14,6 +14,7 @@
                 "4bf5122f344554c53bde2ebb8cd2b7e3d1600ad631c385a5d7cce23c7785459a"
             ]
         },
+        "remembers": [],
         "cached_proof": {
             "targets": [4, 6, 7],
             "hashes": [
@@ -36,6 +37,57 @@
             "0d451c2d9366705a3557fd038c25a40a348cc0effa2dd00dfaaba65749cd0915",
             "7cb7c4547cf2653590d7a9ace60cc623d25148adfbc88a89aeb0ef88da7839ba"
         ],
-        "expected_targets": [4, 6, 7]
+        "expected_targets": [4, 6, 7],
+        "expected_cached_hashes": [
+            "e52d9c508c502347344d8c07ad91cbd6068afc75ff6292f062a09ca381c89e71",
+            "67586e98fad27da0b9968bc039a1ef34c939b9b8e523a8bef89d478608c5ecf6",
+            "ca358758f6d27e6cf45272937977a748fd88391db679ceda7dc7bf1f005ee879"
+        ]
+    },
+    {
+        "update": {
+            "adds": [8],
+            "proof": {
+                "targets": [0, 1, 4],
+                "hashes": [
+                    "e77b9a9ae9e30b0dbdb6f510a264ef9de781501d7b6b92ae89eb059c5ab743db",
+                    "9576f4ade6e9bc3a6458b506ce3e4e890df29cb14cb5d3d887672aef55647a2b",
+                    "34028bbc87000c39476cdc60cf80ca32d579b3a0e2d3f80e0ad8c3739a01aa91"
+                ]
+            },
+            "del_hashes": [
+                "6e340b9cffb37a989ca544e6bb780a2c78901d3fb33738768511a30617afa01d",
+                "4bf5122f344554c53bde2ebb8cd2b7e3d1600ad631c385a5d7cce23c7785459a",
+                "e52d9c508c502347344d8c07ad91cbd6068afc75ff6292f062a09ca381c89e71"
+            ]
+        },
+        "remembers": [0],
+        "cached_proof": {
+            "targets": [4, 6, 7],
+            "hashes": [
+                "e77b9a9ae9e30b0dbdb6f510a264ef9de781501d7b6b92ae89eb059c5ab743db",
+                "df46b17be5f66f0750a4b3efa26d4679db170a72d41eb56c3e4ff75a58c65386"
+            ]
+        },
+        "initial_stump": "Add 8 leaves [0, 1, 2, 3, 4, 5, 6, 7]",
+        "initial_roots": [
+            "b151a956139bb821d4effa34ea95c17560e0135d1e4661fc23cedc3af49dac42"
+        ],
+        "initial_leaves": 8,
+        "cached_hashes": [
+            "e52d9c508c502347344d8c07ad91cbd6068afc75ff6292f062a09ca381c89e71",
+            "67586e98fad27da0b9968bc039a1ef34c939b9b8e523a8bef89d478608c5ecf6",
+            "ca358758f6d27e6cf45272937977a748fd88391db679ceda7dc7bf1f005ee879"
+        ],
+        "expected_roots": [
+            "7f6df6757821144278f5d535261ab4ffc55e2618f7bc62f1a19e7d289061b08b",
+            "6e340b9cffb37a989ca544e6bb780a2c78901d3fb33738768511a30617afa01d"
+        ],
+        "expected_targets": [6, 7, 8],
+        "expected_cached_hashes": [
+                "67586e98fad27da0b9968bc039a1ef34c939b9b8e523a8bef89d478608c5ecf6",
+                "ca358758f6d27e6cf45272937977a748fd88391db679ceda7dc7bf1f005ee879",
+                "6e340b9cffb37a989ca544e6bb780a2c78901d3fb33738768511a30617afa01d"
+        ]
     }  
-]      
\ No newline at end of file
+]
```

diff of the `proof.rs` file to support the test case:
```diff
diff --git a/src/accumulator/proof.rs b/src/accumulator/proof.rs
index 8bac7ba..8e5e70e 100644
--- a/src/accumulator/proof.rs
+++ b/src/accumulator/proof.rs
@@ -568,10 +568,14 @@ mod tests {
             initial_leaves: u64,
             /// The hash of all wallet's utxo
             cached_hashes: Vec<String>,
+            /// The indexes of all the new utxos to cache
+            remembers: Vec<u64>,
             /// After we update our stump, which roots we expect?
             expected_roots: Vec<String>,
             /// After we update the proof, the proof's target should be this
             expected_targets: Vec<u64>,
+            /// After we update the proof, the cached hashes should be this
+            expected_cached_hashes: Vec<String>,
         }
         let contents = std::fs::read_to_string("test_values/cached_proof_tests.json")
             .expect("Something went wrong reading the file");
@@ -632,7 +636,7 @@ mod tests {
                     cached_hashes.clone(),
                     utxos,
                     case_values.update.proof.targets,
-                    vec![],
+                    case_values.remembers,
                     updated.clone(),
                 )
                 .unwrap();
@@ -645,9 +649,16 @@ mod tests {
                 .map(|hash| sha256::Hash::from_hex(hash).unwrap())
                 .collect();
 
+            let expected_cached_hashes: Vec<_> = case_values
+                .expected_cached_hashes
+                .iter()
+                .map(|hash| sha256::Hash::from_hex(hash).unwrap())
+                .collect();
+
             assert_eq!(res, Ok(true));
             assert_eq!(cached_proof.targets, case_values.expected_targets);
             assert_eq!(stump.roots, expected_roots);
+            assert_eq!(cached_hashes, expected_cached_hashes);
         }
     }
     #[test]
```

Reply:
Ok, I think I get this working. I've misinterpreted what `remembers` is, it doesn't says what nodes should be remembered, but what UTXO from the set of new additions in this block should be remembered. I'm also able to return `cached_hashes` from update.   
But I have one question about this test case. You add a leaf with value `8` and uses `6e340b9cffb37a989ca544e6bb780a2c78901d3fb33738768511a30617afa01d` as its hash (the lower root and cached hashes has this value). But `6e34` is the hash of `0`. If I'm not mistaken, `8` hashes to `beead77994cf573341ec17b58bbf7eb34d2711c993c1d976b128b3188dc1829a`. Did I miss something on this `add` value?

Reply:
Ok yeah you're right. 8 hashes to `beead77994cf573341ec17b58bbf7eb34d2711c993c1d976b128b3188dc1829a`.

There was a bug in my go code to generate this. 8 should be `beead77994cf573341ec17b58bbf7eb34d2711c993c1d976b128b3188dc1829a`.